Login First Pass

sammyson101 opened the pull request:
First Pass of login feature

sammyson101 committed:
Login First Pass

mgarber93 commented on src/index.js:
```diff
@@ -1,30 +1,95 @@
+  res.redirect("/");
 });
 
+app.get("/profile", require("connect-ensure-login").ensureLoggedIn(), function (
```
Requires should come at the top of the file, but AFAICT you don't need this package.

```suggestion
 app.get("/profile", passport.authenticate('local'), function (
```

sammyson101 replied:
100% agree, this was just some lazy copy paste

sammyson101 replied:
Moved require to top of file as style dictates. However the suggested change does not work as written and returns a bad request when trying to get profile

mgarber93 commented on src/db/users.js:
```diff
@@ -0,0 +1,32 @@
+var records = [
+  {
+    id: 1,
```
Why start at 1? Find by id is easier if we start at 0.

sammyson101 replied:
For sure

mgarber93 commented on src/db/users.js:
```diff
@@ -0,0 +1,32 @@
+exports.findById = function (id, cb) {
+  process.nextTick(function () {
+    var idx = id - 1;
+    if (records[idx]) {
```
```suggesion
     if (records.hasOwnProperty(idx)) { 
```

sammyson101 replied:
Looking closer this whole thing gets reworked when stating UserID at 0. Updating if statement as suggested and removing the var idx and just using id instead.

mgarber93 replied:
Oh I think this is easier to read than the implicit boolean check. If there was a `null` in the array this would return true now where your way would return false but besides that they are the same.

mgarber93 replied:
I'm curious whats up with the `process.nextTick` though? I normally use promises, but afaict making this logic run immediately would be fine.

mgarber93 commented on src/db/users.js:
```diff
@@ -0,0 +1,32 @@
+    if (records[idx]) {
+      cb(null, records[idx]);
+    } else {
+      cb(new Error("User " + id + " does not exist"));
```
```suggestion
      cb(new Error(`User ${id} does not exist`));
```

mgarber93 commented on src/index.js:
```diff
@@ -1,30 +1,95 @@
+        return cb(err);
+      }
+      if (!user) {
+        return cb(null, false);
```
Should be `null` so the return type is always an object I think.

sammyson101 replied:
So I am def not a Javascript person so I could very much be misunderstanding js, but it was my assumption that everything apart from primitives are objects in JS. So as long as we don't return a primitive type it will always be an examinable object.

mgarber93 replied:
I just meant js lets you return any type from a function, but IMO its easier to reason about if the returned type is always the same. The caller in this case needs to check if it got back an object or boolean in this case.

mgarber93 commented on package.json:
```diff
@@ -34,7 +40,7 @@
   "husky": {
     "hooks": {
-      "pre-commit": "pretty-quick --staged --write src",
+      "pre-commit": "npx pretty-quick --staged --write src",
```
Doesn't seem like this should be required. They don't use npx in the [docs](https://www.npmjs.com/package/pretty-quick#pre-commit-hook)

sammyson101 replied:
So here is where I got this tip from
https://stackoverflow.com/questions/61196309/husky-4-x-not-working-with-visual-studio-git

Although not for VS the issue described is the same issue I was having. That being it can't find lint because its looking for it in C: 

Unsure what to do about this. Is anyone else on windows apart from me?

mgarber93 replied:
It doesn't hurt anything I was just curious.

sammyson101 replied:
I can also 100% keep this local if its only an issue for me and look further into what is going wrong. Really only need this if someone else has an issue

sammyson101 committed:
Update src/db/users.js

Co-authored-by: mgarber93 <mgarber93@gmail.com>

sammyson101 committed:
Update src/index.js

Co-authored-by: mgarber93 <mgarber93@gmail.com>

sammyson101 committed:
Update users.js

sammyson101 committed:
Merge branch 'Login-Feature-1' of https://github.com/mgarber93/hiking-app into Login-Feature-1

sammyson101 committed:
"Pretty" files

mgarber93 commented on package.json:
```diff
@@ -18,7 +18,13 @@
   "license": "ISC",
   "dependencies": {
     "body-parser": "^1.19.0",
+    "connect-ensure-login": "^0.1.1",
```
Lets remove this line and then this looks good to me

sammyson101 replied:
As mentioned above this is module is currently required as written.

mgarber93 replied:
Okay I was able to remove this dependency in 917eb19 do you mind reviewing these changes locally and merging?

mgarber93 replied:
How are you running this btw?  Does running `npm run start:dev` work on your machine?

mgarber93 commented:
As far as directory structure goes I think its normal to lay it out like [this](https://github.com/mgarber93/BitBuyBid/tree/master/server)

sammyson101 committed:
Updates based on PR

Move connect-ensure-login require to top of file

Added a second user and started ID at 0

mgarber93 committed:
Refactor to use passport

mgarber93 committed:
Remove unused dependency
